### PR TITLE
fix(polygamma): apply reflection formula for negative non-integer inputs

### LINKED ISSUE #37635: Incorrect gradient of lax.digamma for negative half-integer inputs
## Description
`jax.lax.digamma` appears to produce an incorrect reverse-mode gradient for negative half-integer inputs.

The derivative of the digamma function is the trigamma function, so `jax.jacrev(lambda x: jax.lax.digamma(x))` should agree with `scipy.special.polygamma(1, x)` for finite non-pole inputs.

The tested values below are negative half-integers, not poles of the digamma function. SciPy returns finite positive trigamma values, but JAX returns small negative gradients.

## Minimal reproducible example:

```
import jax
import jax.numpy as jnp
from scipy.special import polygamma

fn = lambda x: jax.lax.digamma(x)

for v in [-16.5, -19.5, -22.5]:
    a = jnp.array([v])
    grad = float(jax.jacrev(fn)(a)[0, 0])
    expected = float(polygamma(1, v))
    print(f"digamma'({v})  jax={grad:+.4e}  scipy={expected:+.4f}")
```

Actual result:

```
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
digamma'(-16.5)  jax=-5.8807e-02  scipy=+9.8108
digamma'(-19.5)  jax=-4.9990e-02  scipy=+9.8196
digamma'(-22.5)  jax=-4.3471e-02  scipy=+9.8261
```

Expected result:

```
The gradient of jax.lax.digamma should match scipy.special.polygamma(1, x).

Expected values are approximately:
digamma'(-16.5) ~= +9.8108
digamma'(-19.5) ~= +9.8196
digamma'(-22.5) ~= +9.8261
```

### System info (python version, jaxlib version, accelerator, etc.)

```
WARNING:2026-05-07 00:33:38,915:jax._src.xla_bridge:794: An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
jax:    0.6.2
jaxlib: 0.6.2
numpy:  2.2.6
python: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0]
device info: cpu-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='MSI', release='5.15.167.4-microsoft-standard-WSL2', version='#1 SMP Tue Nov 5 00:21:55 UTC 2024', machine='x86_64')

nvidia-smi:
NVIDIA-SMI 545.23.06
Driver Version: 545.92
CUDA Version: 12.3
GPU: NVIDIA GeForce RTX 3080, 8192 MiB

Note: Reproduced on CPU fallback because CUDA-enabled jaxlib is not installed.
```

### PR DESCRIPTION
**Pull request by Sumu004**
## Problem

`Polygamma(n, x)` for `n >= 1` calls `Zeta(n+1, x)` directly. XLA's `Zeta` implementation uses the Euler-Maclaurin expansion (N=9 initial terms, M=12 correction terms) which requires `q > 0` to converge. For negative `x`, the expansion diverges, producing catastrophically wrong results:

```
polygamma(1, -9.5)   →  -2 854 289 648 037  (correct: 9.770)
polygamma(1, -16.5)  →  -0.059              (correct: 9.811)
polygamma(2, -8.5)   →  893 286             (correct: -0.012)
```

This also corrupts `grad(digamma)` in JAX (which computes `polygamma(1, x)`) and any framework building on `chlo.polygamma`.

Reported in jax-ml/jax#37635 and tracked in the JAX numerical accuracy issue jax-ml/jax#27088.

Note: `Digamma` already applies its own reflection formula for `x < 0.5` (see line 679). This fix brings `Polygamma` for `n >= 1` to parity.

## Fix

For `n ∈ {1, 2, 3, 4}` and `x < 0` (non-integer), apply the reflection formula derived by differentiating the digamma reflection identity `ψ(1-x) − ψ(x) = π cot(πx)` exactly `n` times:

```
ψ^(n)(x) = (−1)^n ψ^(n)(1−x) − gₙ(x),   x ∉ ℤ
```

Since `1−x > 0` for `x < 0`, `Zeta(n+1, 1−x)` evaluates correctly. The closed-form trig terms `gₙ(x)` are:

| n | gₙ(x) |
|---|---|
| 1 | `−π²/sin²(πx)` |
| 2 | `2π³ cos(πx)/sin³(πx)` |
| 3 | `−2π⁴(2 + cos(2πx))/sin⁴(πx)` |
| 4 | `8π⁵ cos(πx)(2 + cos²(πx))/sin⁵(πx)` |

For `n = 0` the existing `Digamma(x)` path already handles negative `x`.  
For `n ≥ 5` the existing (inaccurate) direct path is preserved — those orders are rarely needed.

To improve precision near integer `x` (where `π·x` has cancellation), the trig arguments use `reduced_x = x + |floor(x + 0.5)|` ∈ [−0.5, 0.5], following the same technique already used in `Digamma`. All `gₙ` formulas are periodic in `x` with period 1, so `gₙ(reduced_x) = gₙ(x)`.

## Before / After

```
polygamma(1, -9.5)   before: -2 854 289 648 037    after: 9.770  ✓
polygamma(1, -16.5)  before: -0.059                after: 9.811  ✓
polygamma(3, -0.5)   before: wrong (n=3 path)      after: 193.41 ✓
```

Fixes jax-ml/jax#37635

### CODE REVIEW

**Review by dimitar-asenov — CHANGES_REQUESTED** [review]
Thank you for implementing a fix for this. Please add a test that verifies the fix.

**Comment by Sumu004** [comment]
Done — added `TEST_F(MathTest, PolygammaNegativeInputs)` in the latest commit covering:

- **n = 1 (trigamma)** at `x ∈ {-0.5, -3.3, -8.5, -9.5, -16.5}` — these were the catastrophically wrong cases (`-2.85e12` instead of `9.77`)
- **n = 2 (tetragamma)** at `x ∈ {-0.5, -8.5}`
- **n = 3 (pentagamma)** at `x ∈ {-0.5, -3.3}`
- **Positive inputs** (`x ∈ {0.5, 2.0}`, n = 1 and 2) — regression check that the fix does not affect the correct path

Golden values generated by `scipy.special.polygamma`. Float32 tolerance uses `ErrorSpec(1e-2, 1e-4)` for the reflected path (one extra trig evaluation vs. direct); positive inputs use the existing `kErrorSpec`.

**Comment by Sumu004** [comment]
@dimitar-asenov tests were added in the second commit (d9c0694e) — sorry for the confusion on timing. `TEST_F(MathTest, PolygammaNegativeInputs)` covers:
- **n=1 (trigamma)** at x ∈ {-0.5, -3.3, -8.5, -9.5, -16.5} — the catastrophically wrong cases
- **n=2 (tetragamma)** at x ∈ {-0.5, -8.5}
- **n=3 (pentagamma)** at x ∈ {-0.5, -3.3}
- **Positive inputs** — regression check that the fix does not affect the correct path

Golden values generated by `scipy.special.polygamma`. Happy to adjust tolerance or add more cases if needed.

**Review by dimitar-asenov — APPROVED** [review]

**Comment by dimitar-asenov** [comment]
Thank you.

**Comment by Sumu004** [comment]
Hi, this PR has been approved by @dimitar-asenov. Could a maintainer approve the pending workflows to unblock CI and trigger the copybara sync? Happy to address any further feedback. Thanks!

**Comment by dimitar-asenov** [comment]
@Sumu004 There is a failing test after this change:

`hlo/builder/tests:math_polygamma.hlo.test`

```
[xla/hlo/builder/tests/math_polygamma.hlo:9]
error: CHECK-NEXT: is not on the line after the previous match
// CHECK-NEXT: %[[arg0_1:[^ ]+]] = f32[] parameter(0)
               ^
<stdin>:8:2: note: 'next' match was here
 %arg0.1 = f32[] parameter(0)
 ^
<stdin>:5:35: note: previous match ended here
 %constant.286 = f32[] constant(1)
                                  ^
<stdin>:6:1: note: non-matching line after previous match is here
 %constant.392 = f32[] constant(0)
^

Input file: <stdin>
Check file: xla/hlo/builder/tests/math_polygamma.hlo

-dump-input=help explains the following input dump.

Input was:
<<<<<<
        1: HloModule main, entry_computation_layout={(f32[], f32[])->f32[]} 
        2:  
        3: %xla_builder.math.Polygamma.1 (arg0.1: f32[], arg1.1: f32[]) -> f32[] { 
        4:  %constant.272 = f32[] constant(0) 
        5:  %constant.286 = f32[] constant(1) 
        6:  %constant.392 = f32[] constant(0) 
        7:  %constant.406 = f32[] constant(1) 
        8:  %arg0.1 = f32[] parameter(0) 
next:9      !~~~~~~~~~~~~~~~~~~~~~~~~~~~  error: match on wrong line
        9:  %floor.25 = f32[] floor(%arg0.1) 
       10:  %compare.72 = pred[] compare(%arg0.1, %floor.25), direction=NE 
       11:  %constant.454 = f32[] constant(0) 
       12:  %compare.73 = pred[] compare(%arg0.1, %constant.454), direction=LT 
       13:  %or.1 = pred[] or(%compare.72, %compare.73) 
        .
        .
        .
>>>>>>

--

********************
```

Could you please make sure this passes. Thanks.

**Comment by Sumu004** [comment]
@dimitar-asenov Thanks for catching that — you're right, the `math_polygamma.hlo` golden file is now stale. The reflection-formula change adds a fair number of new ops (`Sin`, `Cos`, extra constants, a `Select` branch), so essentially the whole autogenerated `CHECK` block needs to be refreshed, not just a line or two.

I don't currently have a full XLA/Bazel build environment set up locally (building `hlo-opt` from scratch pulls in LLVM/MLIR and needs significantly more disk/compute than I have on hand right now), so regenerating the exact instruction numbering myself isn't practical without a multi-hour build.

Would you be able to run the regeneration script against your build and push the updated golden file? It should just be:

```
generate_hlo_test_checks.py xla/hlo/builder/tests/math_polygamma.hlo -i -- hlo-opt --passes=test-only-xla-builder --split-input-file {} | FileCheck {}
```

Happy to review the diff afterwards to confirm it matches the expected lowering for the reflection path. Sorry for the extra round trip, and thanks again for the review!
